File: fsharp4csharpDevs/fsharpDevs/2_Collections.fs
﻿module Collections

//Arrays

//this is an array
let arr = [|1;2;3|]

//same as this
let arr2 = 
    [|
        1
        2
        3
    |]
    
//array of numb from 0 to 99
let arrOfNumb = [|0..99|]

//array with steps
let arrOfNumb2 = [|0..5..99|]

//Initialized from loop
let arrOfNumb3 = [|for i in 0..5..99 do yield i * i|]


//This is a function that returns a array 
let RandomFruits count = 
    let r = System.Random()
    let fruits = [|"apple";"orange";"pear"|]
    [|
        for i in 1..count do
        let index = r.Next(3);
        yield fruits.[index]
    |]
    
    
//This is a function that returns a array different way
//This One is not == to C# function RandomFruitsEnumerable!!!!!
let RandomFruits2 count = 
    let r = System.Random()
    let fruits = [|"apple";"orange";"pear"|]
    Array.init count (fun iteratorNo ->
        let index = r.Next(3)
        fruits.[index]
    )


//Access an element of index 0
//arr.[0]


//Array Iterating
let IteratingArray arrNumb =
    for num in arrNumb do
        printfn "i am number: %i" num

//Filtering Arrays
//What im gonna filter:
let arrNumb = [|for i in 0..5..99 do yield i * i|]

//base on condition
let IsEven n = n % 2 = 0

//filter results
let filterResults = Array.filter (fun x -> IsEven x) arrNumb

//Sorting filtering and Iterating in one go
let PrintLongWords(words: string[]) =
    words
    |> Array.filter(fun w -> w.Length > 8)
    |> Array.sort
    |> Array.iter(fun w -> printfn w)

File: fsharp4csharpDevs/csharpDevs/2_Collections.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace csharpDevs
{
    class Collections
    {
        //this is an array
        int[] arr = new int[] { 1, 2, 3 };

        //same as this
        int[] arr2 = new int[] 
        {
            1,
            2,
            3
        };

        //array of numb from 0 to 99
        int[] arrOfNumb = GetArray_0_To_99();

        private static int[] GetArray_0_To_99()
        {
            int[] arr = new int[100];
            for (int i = 0; i < arr.Length; i++)
            {
                arr[i] = i;
            }
            return arr;
        }

        //array with steps
        int[] arrOfNumb2 = GetArrayOfSteps();

        private static int[] GetArrayOfSteps()
        {
            int[] arr = new int[20];
            for (int i = 0; i < arr.Length; i++)
            {
                arr[i] = i*5;
            }
            return arr;
        }


        //Initialized from loop
        int[] arrOfNumb3 = InitializedFromLoop();

        private static int[] InitializedFromLoop()
        {
            int[] arr = new int[20];
            for (int i = 0; i < arr.Length; i++)
            {
                arr[i] = i * i * 5;
            }
            return arr;
        }

        //This is a function that returns a array
        public string[] RandomFruits(int count)
        {
            string[] result = new string[count];
            string[] fruits = { "apple", "orange", "pear" };
            var r = new System.Random();

            for (int i = 0; i < count; i++)
            {
                int index = r.Next(3);
                result[i] = fruits[index];
            }

            return result;
        }

        //This is a function that returns a array as IEnumerable
        public IEnumerable<string> RandomFruitsEnumerable(int count)
        {
            string[] fruits = { "apple", "orange", "pear" };
            var r = new System.Random();

            for (int i = 0; i < count; i++)
            {
                yield return fruits[r.Next(3)];
            }
        }


        //Access an element of index 0
        //arr[0]

        //Array Iterating
        public void IteratingArray(int[] arrNumb)
        {
            foreach (var num in arrNumb)
            {
                Console.WriteLine("i am number: %i", num);
            }
        }

        //Filtering Arrays
        //What im gonna filter:

        //base on condition
        static bool IsEven(int n) { return n % 2 == 0; }

        //filter results
        int[] filterResults = GetArray_0_To_99().Where(x => IsEven(x)).ToArray();

        //Sorting filtering and Iterating in one go
        public void PrintLongWords(string[] words)
        {
            var arr = words.Where(w => w.Length > 8).ToArray());
            Array.Sort(arr);
            foreach (var arg in arr)
            {
                Console.WriteLine(arg);
            }
        }
    }
}
